2099: Redirect problems

laurent-bientz opened the pull request:
#2099, I've change the redirect's workflow after submitting the new, edit or delete form:
* if we can redirect to List > go on List
* else if from new or edit and we can redirect to edit > go on Edit
* else if from new and we can redirect to add > go on Add
* else > go on default backend homepage

so:
* when possible, redirect to List
* if trying to edit but list disabled, put flash msg + redirect to Edit (on the current recordset)
* if trying to add but list disabled, put flash msg + redirect to Edit (on the recordset created)
* if trying to add but list + edit disabled, put flash msg + redirect to Add (empty form)
* if trying to delete but list disabled, redirect to default backend

ps: I've only added translations for flash messages in french (native) and english.

javiereguiluz commented on src/Controller/AdminController.php:
```diff
@@ -820,18 +820,53 @@ public function renderCssAction()
-            return $this->redirect(urldecode($referrerUrl));
+        // redirect on list if possible
+        if ($this->isActionAllowed('list')) {
+            return $this->redirect($this->generateUrl('easyadmin', array(
```
Instead of `redirect()` we can use the shortcut `redirectToRoute()`

laurent-bientz replied:
👍

javiereguiluz commented on src/Controller/AdminController.php:
```diff
@@ -820,18 +820,53 @@ public function renderCssAction()
-        if (!empty($referrerUrl)) {
-            return $this->redirect(urldecode($referrerUrl));
+        // redirect on list if possible
+        if ($this->isActionAllowed('list')) {
```
In case `list` is allowed, we can redirect directly to the referrer URL instead of building it again, yes?

laurent-bientz replied:
Can u confirm that referer param is always set when we're coming from List?

javiereguiluz commented on src/Controller/AdminController.php:
```diff
@@ -820,18 +820,53 @@ public function renderCssAction()
 
-        if ($this->isActionAllowed('list')) {
+        // else from new|edit action, redirect on edit if possible
+        elseif (in_array($refererAction, array('new', 'edit')) && $this->isActionAllowed('edit')) {
```
We recently fixed a bug where custom query params weren't saved. People often add custom params and now we maintain them continuously, from the list view, to the actions shown in list, the buttons of the edit/new views and back to the original list view. Moreover, we use certain params to store the sorting and with this code that would be lost.

I propose you the following. Let's extract the query string from the referrer URL and let's overwrite just the `action` query string param to redirect to `edit` or `new`.

laurent-bientz replied:
I understand your point but in my opinion, there can not be any referrer param in case of disabled listing or I miss something?

In case of redirects:
- edit > edit
- add > edit
- add > add

User access the source action by clicking directly on the menu, so no possible referer?

javiereguiluz commented on src/Resources/translations/EasyAdminBundle.en.xlf:
```diff
@@ -181,6 +181,16 @@
+            <!-- Flash messages -->
+            <trans-unit id="flash.new.success">
+                <source>flash.new.success</source>
+                <target>Your element has been created.</target>
```
The policy of this bundle is to never display messages when things work (because that's the normal case); only when they fail. So let's remove this confirmation messages. By the way, the workflow you designed makes a lot of sense, so redirects will always be common sense and the user won't be lost.

laurent-bientz replied:
You're the boss  :stuck_out_tongue: even if I'm not rly fan about that.

laurent-bientz commented:
@javiereguiluz 
- I added a redirect to referrer if not empty before trying to regenerate list route
- I updated to use the redirectToRoute shortcut instead of redirect
- For me, it can't be possible to have a referer param if list is disabled, if I'm wrong, plz explain to me how it can be possible?
- I remove flash msgs and translations  :sweat:

javiereguiluz commented on src/Controller/AdminController.php:
```diff
@@ -820,16 +820,44 @@ public function renderCssAction()
+                'menuIndex' => $this->request->query->get('menuIndex'),
+                'submenuIndex' => $this->request->query->get('submenuIndex'),
+                'id' => ('new' === $refererAction)
+                    ? PropertyAccess::createPropertyAccessor()->getValue($this->request->attributes->get('easyadmin')['item'], $this->entity['primary_key_field_name'])
```
My only remaining question about this is: why do we need to add an `id` query param in the case of `new`? Can't we just make this?: `'id' => $this->request->query->get('id')` In `new`, the `id` query attribute will be `null` and that's OK in this case, right?

laurent-bientz replied:
Cause in my opinion, I think it's better to be redirected on the recordset we just created after add (if edit is not disabled), that's why I use the PropertyAccess to get the value from the PK.

Your case is just below (lines 853-861), edit is disabled and we're on add, so we redirect on add without id.

What do u think about that?

javiereguiluz replied:
I understood your code wrongly. This looks OK to me!

javiereguiluz commented:
@laurent-bientz thanks for this bug fixing! Even if you have participated in lots of issues in this repo, this was your first code contribution, so thanks a lot! I'm going to add tests for this in a separate PR.

laurent-bientz commented:
hi @ossinkine,

cause Javier asked me to do it ;)
https://github.com/EasyCorp/EasyAdminBundle/pull/2150/files/b72574bad9ebc32564d719b55209b20f2e85fea5

Feel free to discuss about it with him.

ossinkine commented:
@javiereguiluz Could you please explain how is the referrer URL related to the list action?
